src/os/bluestore/bluefs_types.cc

```diff
@@ -242,7 +255,16 @@ void bluefs_transaction_t::bound_encode(size_t &s) const {
 void bluefs_transaction_t::encode(bufferlist& bl) const
 {
   uint32_t crc = op_bl.crc32c(-1);
-  ENCODE_START(1, 1, bl);
+  __u8 version = 1;
+  __u8 compat = 1;
+  if (wal_v2_count > 0) {
```

Comment:
If we have no op_dir_link in a transaction, then it will revert to v1.
Even though fnodes updated in it might be wal_v2.
Its creepy.

I think the selector for version 1 vs version 2 should apply to ALL bluefs_transaction_t objects; with a compact_log being the place that is able to switch v2->v1.